Sources/Fuzzilli/Corpus/MarkovCorpus.swift

```diff
@@ -0,0 +1,196 @@
+        }
+        assert(prog!.size != 0)
+        return prog!
+    }
+
+    /// For the first 500 executions, randomly choose a program. This is done to build a base list of edge counts
+    /// Once that base is acquired, provide samples that trigger an infrequently hit edge
+    public func randomElementForMutating() -> Program {
+        totalExecs += 1
+        if(size <= 1){
+            return genericProg!
+        }
+        // Only do computationally expensive work choosing the next program when there is a solid
+        // baseline of execution data. The data tracked in the statistics module is not used, as modules are intended 
+        // to not be required for the fuzzer to function.
+        if totalExecs > 1000 {
+            // Check if more programs are needed
+            if programExecutionQueue.count == 0 {
+                regenProgramList()
+            }
+            assert(programExecutionQueue.count > 0)
+            var prog : Program
+            if let tempProg = currentProg, remainingEnergy > 0 {
+                prog = tempProg
+                remainingEnergy -= 1
+            } else {
+                prog = programExecutionQueue.popLast()!
+                currentProg = prog
+                remainingEnergy = energyBase()
+            }
+            return prog
+        } else {
+            return allIncludedPrograms.randomElement()!
+        }
+    }
+
+    private func regenProgramList() {
+        assert(programExecutionQueue.count == 0)
+        let covEdgesBuffPtr = covEvaluator.getEdgeCountPtr()!
+        // Per https://developer.apple.com/documentation/swift/unsafebufferpointer, making an array from an UnsafeBufferPtr copies the memory
+        var covEdgesArr = Array(covEdgesBuffPtr)
+        covEdgesArr.sort()
+
+        // Find the edge with the smallest count
+        var startIndex = -1
+        for (i, val) in covEdgesArr.enumerated() {
+            if startIndex == -1 && val != 0 {
+                startIndex = i
+                break
+            }
+        }
+        assert(startIndex != -1)
+        
+        // Find the nth interesting edge's count
+        let desiredEdgeCount = max(size / 8, 30)
+        let endIndex = min(startIndex + desiredEdgeCount, covEdgesArr.count - 1)
+        let maxEdgeCountToFind = covEdgesArr[endIndex]
+        let amountToAge = energyBase() * numConsecutiveMutations // Likely overaggressive
+
+        // Find the n edges with counts <= maxEdgeCountToFind. Age them appropriately,
+        // to ensure non-deterministic samples are not continuously selected
+        for (i, val) in covEdgesBuffPtr.enumerated() {
+            // Applies dropout on otherwise valid samples, to ensure variety between instances
+            // This will likely select some samples multiple times, which is acceptable as
+            // it is proportional to how many infrquently hit edges the sample has
+            if val != 0 && val <= maxEdgeCountToFind && probability(0.25){ 
```

Comment:
This implies the dropout ratio is 75%, should it really be that high? Can the ratio be a class constant?
Also, if an edge is dropped, shouldn't another take it's spot instead? IIUC, the way it is now, it will lead to the programExecutionQueue containing less elements in the end

Reply:
I've changed the effective ratio to 10%, and made it a cli flag.

My preference is to not have another sample take its slot. Regenerating the program list is fast and infrequent enough that it isn't a significant performance hit to miss out on some programs, with some reduced code complexity. WDYT?

Reply:
Sounds good to me. Maybe just change it to something like `&& (probability(P) || programExecutionQueue.isEmpty) {` to guarantee that the queue will contain at least one element at the end though?